x-pack/plugins/apm/server/lib/fleet/source_maps.ts

```diff
@@ -0,0 +1,166 @@
+      const body = await doUnzip(Buffer.from(artifact.body, 'base64'));
+      return {
+        ...artifact,
+        body: JSON.parse(body.toString()) as ApmArtifactBody,
+      };
+    })
+  );
+}
+
+function getApmArtifactClient(fleetPluginStart: FleetPluginStart) {
+  return fleetPluginStart.createArtifactsClient('apm');
+}
+
+export async function listArtifacts({
+  fleetPluginStart,
+}: {
+  fleetPluginStart: FleetPluginStart;
+}) {
+  const apmArtifactClient = getApmArtifactClient(fleetPluginStart);
+  const artifacts = await apmArtifactClient.listArtifacts({
+    kuery: 'type: sourcemap',
+  });
+
+  return decodeArtifacts(artifacts.items);
+}
+
+export async function createApmArtifact({
+  apmArtifactBody,
+  fleetPluginStart,
+}: {
+  apmArtifactBody: ApmArtifactBody;
+  fleetPluginStart: FleetPluginStart;
+}) {
+  const apmArtifactClient = getApmArtifactClient(fleetPluginStart);
+  const identifier = `${apmArtifactBody.serviceName}-${apmArtifactBody.serviceVersion}`;
+
+  return apmArtifactClient.createArtifact({
+    type: 'sourcemap',
+    identifier,
+    content: JSON.stringify(apmArtifactBody),
+  });
+}
+
+export async function deleteApmArtifact({
+  id,
+  fleetPluginStart,
+}: {
+  id: string;
+  fleetPluginStart: FleetPluginStart;
+}) {
+  const apmArtifactClient = getApmArtifactClient(fleetPluginStart);
+  return apmArtifactClient.deleteArtifact(id);
+}
+
+export function getPackagePolicyWithSourceMap({
+  packagePolicy,
+  artifacts,
+}: {
+  packagePolicy: PackagePolicy;
+  artifacts: ArtifactSourceMap[];
+}) {
+  if (!artifacts.length) {
+    return packagePolicy;
+  }
+  const [firstInput, ...restInputs] = packagePolicy.inputs;
+  return {
+    ...packagePolicy,
+    inputs: [
+      {
+        ...firstInput,
+        config: {
+          ...firstInput.config,
+          [APM_SERVER]: {
+            value: {
+              ...firstInput?.config?.[APM_SERVER].value,
+              rum: {
+                source_mapping: {
+                  metadata: artifacts.map((artifact) => ({
+                    'service.name': artifact.body.serviceName,
+                    'service.version': artifact.body.serviceVersion,
+                    'bundle.filepath': artifact.body.bundleFilepath,
+                    'sourcemap.url': artifact.relative_url,
+                  })),
+                },
+              },
+            },
+          },
+        },
+      },
+      ...restInputs,
+    ],
+  };
+}
+
+export async function updateSourceMapsOnFleetPolicies({
+  fleetPluginStart,
+  savedObjectsClient,
+  elasticsearchClient,
+}: {
+  fleetPluginStart: FleetPluginStart;
+  savedObjectsClient: SavedObjectsClientContract;
+  elasticsearchClient: ElasticsearchClient;
+}) {
+  const artifacts = await listArtifacts({ fleetPluginStart });
+
+  const apmFleetPolicies = await fleetPluginStart.packagePolicyService.list(
+    savedObjectsClient,
+    { kuery: 'ingest-package-policies.package.name:apm' }
```

Comment:
This should be a shared constant since it's also used here: https://github.com/elastic/kibana/blob/master/x-pack/plugins/apm/server/lib/fleet/get_apm_package_policies.ts#L28

Reply:
There's also some opportunity to abstract the updating of all policy matches in parallel as shared logic (also seen here: https://github.com/elastic/kibana/blob/master/x-pack/plugins/apm/server/lib/fleet/sync_agent_configs_to_apm_package_policies.ts#L44-L58), but we can also wait on these code organization tasks until more fleet PRs are merged.

Reply:
I'm now calling `getApmPackgePolicies` from https://github.com/elastic/kibana/blob/master/x-pack/plugins/apm/server/lib/fleet/get_apm_package_policies.ts#L16:L16